python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -153,15 +176,86 @@ def caller(data, index, size):
     call_cuda_functions[funcname.lower()][type_key] = caller
 
 
-def _make_unary_attr(funcname):
-    class GroupUnaryReductionAttrTyping(AbstractTemplate):
-        key = f"GroupType.{funcname}"
+class GroupOpBase(AbstractTemplate):
+    def make_error_string(self, args):
+        fname = self.key.__name__
+        sr_err = ", ".join(["Series" for _ in range(len(args))])
+        return (
+            f"{fname}({sr_err}) is not supported by JIT GroupBy "
+            f"apply. Supported features are listed at:\n{_UDF_DOC_URL}"
+        )
+
+    def generic(self, args, kws):
+        # early exit to make sure typing doesn't fail for normal
+        # non-group ops
+        if not all(isinstance(arg, GroupType) for arg in args):
+            return None
+        # check if any groups are poisoned for this op
+        for arg in args:
+            if isinstance(arg.group_scalar_type, types.Poison):
+                raise UDFError(
+                    f"Use of a column of {arg.group_scalar_type.ty} detected "
+                    "within UDF body. \nOnly columns of the following dtypes "
+                    "may be used through the \nGroupBy.apply() JIT engine: "
+                    f"{[str(x) for x in SUPPORTED_GROUPBY_NUMPY_TYPES]}"
+                )
```

Comment:
Rather than erroring on the first encountered error, maybe collect all errors and then raise with the full set in case there are more than one? Saves the user having to iteratively remove parts of the UDF.

Reply:
That's a great suggestion and I hadn't thought about it from that perspective. In fact, normally this is what numba does by default. It goes through these typing classes `generic` functions one by one and if none of them return a signature (signaling success) it raises a general `TypingError`. This usually prints the big numba error string we're used to seeing, and contained within it are any actual errors raised by the `generic` functions themselves.

The thing is that this PR was kind of designed to break us out of that mechanism as convenient as it is for what you're suggesting - that is, raising and torpedoing typing phase _immediately_ with the goal of communicating something minimal and clear to the user instead. It seems like its worth exploring going back to an approach that mixes the existing mechanism with this new one and maybe just cleans up the default numba string, but I'd probably need to dive a little deeper into the numba internals next release to explore that in a way that bears fruit.